java/main/jglome/Glome.java

```diff
@@ -0,0 +1,357 @@
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          "The received tag has invalid length. Expected " + minPeerTagLength + " or more, got "
+              + peerTag.length + ".");
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Common part of the constructors. Initialises peer's public key, minimum length of a peer's tag,
+   * base point for Curve25519 curve. Calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum length of a peer's tag.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  private void initMacKeys(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    this.peerKey = peerKey;
+    if (minPeerTagLength < MIN_TAG_LENGTH || minPeerTagLength > MAX_TAG_LENGTH) {
+      throw new MinPeerTagLengthOutOfBoundsException(
+          "minPeerTagLength argument should be in [1..32] range. Got " + minPeerTagLength + ".");
+    }
+    this.minPeerTagLength = minPeerTagLength;
+
+    KeyAgreement ka;
+    try {
+      ka = KeyAgreement.getInstance("XDH");
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    try {
+      ka.init(userPrivateKey());
+      ka.doPhase(peerKey, true);
+    } catch (InvalidKeyException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    byte[] sharedSecret = ka.generateSecret();
+    byte[] userMac = subjectPublicKey(userPublicKey());
+    byte[] peerMac = subjectPublicKey(peerKey);
+
+    this.userMacKey = getMacKey(sharedSecret, peerMac, userMac);
+    this.peerMacKey = getMacKey(sharedSecret, userMac, peerMac);
+  }
+
+  /**
+   * @param pk public key.
+   * @return byte representation of {@code pk}.
+   */
+  private byte[] subjectPublicKey(XECPublicKey pk) {
+    byte[] encodedPK = pk.getEncoded();
+
+    return Arrays.copyOfRange(encodedPK, encodedPK.length - 32, encodedPK.length);
+  }
+
+  private XECPublicKey derivePublicKeyFromPrivate(XECPrivateKey privateKey) {
+    KeyPairGenerator keyPairGenerator;
+    try {
+      keyPairGenerator = KeyPairGenerator.getInstance("X25519");
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    try {
+      keyPairGenerator.initialize(X25519, new StaticSecureRandom(privateKey.getScalar()));
```

Comment:
I'd have expected this to just seed the RNG. Does this work to actually derive a key? :o